vllm/v1/worker/gpu_model_runner.py

```diff
@@ -355,6 +362,51 @@ def _prepare_inputs(self, scheduler_output: "SchedulerOutput"):
             self.device, non_blocking=True)
         slot_mapping = self.slot_mapping_cpu[:total_num_scheduled_tokens].to(
             self.device, non_blocking=True).long()
+
+        # Prepare for cascade attention if needed.
+        common_prefix_len = (scheduler_output.num_common_prefix_blocks *
+                             self.block_size)
+        if common_prefix_len == 0:
+            # Common case.
+            use_cascade = False
+        else:
+            # The common prefix should be already computed and stored in KV
+            # cache before this step.
```

Comment:
Is this true for the first batch?

Reply:
Yes, this is required by the first kernel (computing prefix) in cascade attention.

For the first kernel, we treat all the query tokens (from different sequences) as a single sequence, and use bi-directional attention (causal=False) to compute the partial output. Therefore, every token in the query attends to the prefix without any masking. If the input query overlaps with the shared prefix in KV cache, it means we need to consider masking and can't use the bi-directional attention kernel.

Reply:
Actually, while writing this comment, I realized that we can relax the constraint a little bit: If the input query overlaps with the shared prefix in KV cache **only by 1 token**, the first kernel (computing prefix) is enough. Let me quickly prototype this.

Reply:
> if the input query overlaps with the shared prefix in KV cache only by 1 token, the first kernel (computing prefix) is enough.

After second thought, I found that this is a pretty niche case that probably isn't worth optimizing. The described case is essentially when all requests in the batch have the same query (of len 1) & same context in KV cache, which in turn means that the requests are duplicated. I think this is a very rare case.

Does this make sense to you? @comaniac